src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
+            prevNode.next = newNode;
+        } else if (prevNode == null || prevNode.prev == null) {
+            head = newNode;
+        }
+        nextNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexNotAddValidator(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexNotAddValidator(index);
+        Node<T> indexNode = findNodeByIndex(index);
+        T oldValue = indexNode.value;
+        indexNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexNotAddValidator(index);
+        Node<T> deletedNode = findNodeByIndex(index);
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else {
+            Node<T> prevNode = (index > 0) ? deletedNode.prev : null;
+            Node<T> nextNode = (index < size) ? deletedNode.next : null;
+            if (index == 0) {
+                nextNode.prev = null;
+                head = nextNode;
+            } else if (index == size - 1) {
+                prevNode.next = null;
+                tail = prevNode;
+            } else {
+                prevNode.next = deletedNode.next;
+                nextNode.prev = deletedNode.prev;
+            }
+        }
+        size--;
+        return deletedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        int index = 0;
+        do {
+            if (object == current.value
+                    || (current.value != null && current.value.equals(object))) {
+                remove(index);
+                return true;
+            }
+            index++;
+            current = current.next;
+        } while (current.next != null);
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void indexAddValidator(int index) {
+        if (index > size) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
+        } else if (index < 0) {
+            throw new IndexOutOfBoundsException(NEGATIVE_INDEX_MESSAGE);
+        }
+    }
+
+    private void indexNotAddValidator(int index) {
+        if (index >= size) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
+        } else if (index < 0) {
+            throw new IndexOutOfBoundsException(NEGATIVE_INDEX_MESSAGE);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
```

Comment:
Added null checks in the findNodeByIndex(int index) method to handle cases where the linked list is empty.